### PR TITLE
Add QueryBuilder::insertIfNotExists()

### PR DESCRIPTION
**Pull request by nickvergessen**
Fixes https://github.com/owncloud/core/pull/20018/files#r48609250

@icewind1991 @DeepDiver1975 @rullzer

### CODE REVIEW

File: lib/private/db/querybuilder/querybuilder.php [hunk 1]
```diff
@@ -1104,4 +1104,21 @@ public function getColumnName($column, $tableAlias = '') {
...
+	 * Uses {@see Connection::executeUpdate}
+	 *
+	 * @param string $table
+	 * @param array $insert data that should be inserted into the table  (column name => value)
+	 * @param array|null $compare List of values that should be checked for "if not exists"
+	 *				If this is null or an empty array, all keys of $insert will be compared
+	 *				Please note: text fields (clob) must not be used in the compare array
+	 * @return int number of inserted rows
+	 */
+	public function insertIfNotExists($table, array $insert, array $compare = null) {
```
Comment (DeepDiver1975) [inline lib/private/db/querybuilder/querybuilder.php]: do we really what this on the querybuilder? it's not really building a query ....

Comment (nickvergessen) [inline lib/private/db/querybuilder/querybuilder.php]: Yeah not sure either, but it allows to "forget" the prefix and you can use lastInsertId() of the qb without table name afterwards etc.

Comment (DeepDiver1975) [inline lib/private/db/querybuilder/querybuilder.php]: yes - but if we want to fix this it's worth a second method in the connection - not in the builder from my pov

**Comment by MorrisJobke** [comment]
Close in favor of #21714?

File: lib/private/db/querybuilder/querybuilder.php (hunk 1 above)
Comment (icewind1991) [inline lib/private/db/querybuilder/querybuilder.php]: Should be easy to fix the existing `insertIfNotExists` to handle prefixing if `'*PREFIX*'` isn't passed as part of the table name

Comment (nickvergessen) [inline lib/private/db/querybuilder/querybuilder.php]: yeah, but that breaks existing usages, that don't use the prefix ;)
So we can't just introduce this. anyway, lets not waste too much time on this.

**Pull request closed without merging**